src/world/scheduler/builder.rs

```diff
@@ -810,6 +810,48 @@ fn multiple_mixed() {
     assert_eq!(scheduler.default, "Systems");
 }
 
+#[test]
+fn append_optimizes_batches() {
+    use crate::{View, ViewMut, World};
+
+    fn systemA1(_: ViewMut<'_, usize>, _: ViewMut<'_, u32>) {}
+    fn systemA2(_: View<'_, usize>, _: ViewMut<'_, u32>) {}
+    fn systemB1(_: View<'_, usize>) {}
+
+    let world = World::new();
+
+    let mut group_a = Workload::builder("Group A");
+    group_a
+        .try_with_system((|world: &World| world.try_run(systemA1), systemA1))
+        .unwrap()
+        .try_with_system((|world: &World| world.try_run(systemA2), systemA2))
+        .unwrap();
+    let mut group_b = Workload::builder("Group B");
+    group_b
+        .try_with_system((|world: &World| world.try_run(systemB1), systemB1))
+        .unwrap();
+
+    let combined_info = Workload::builder("Combined")
+        .append(&mut group_a)
+        .append(&mut group_b)
+        .add_to_world_with_info(&world)
+        .unwrap();
+
+    assert_eq!(
+        combined_info.batch_info.len(),
+        2,
+        "systemB1 should move into second batch with systemA2"
+    );
+    let scheduler = world.scheduler.try_borrow_mut().unwrap();
+    assert_eq!(scheduler.systems.len(), 3);
+    assert_eq!(scheduler.workloads.len(), 1);
+    assert_eq!(
+        scheduler.workloads.get("Combined"),
+        Some(&vec![vec![0], vec![1, 2]])
+    );
+    assert_eq!(scheduler.default, "Combined");
+}
+
 #[test]
```

Comment:
BTW, I was playing with the optimizer a bit more and I wrote the following failing test:

```rust
#[test]
fn append_ensures_multiple_batches_can_be_optimized_over() {
    use crate::{View, ViewMut, World};

    fn systemA1(_: ViewMut<'_, usize>, _: ViewMut<'_, u32>) {}
    fn systemA2(_: View<'_, usize>, _: ViewMut<'_, u32>) {}
    fn systemB1(_: View<'_, usize>) {}
    fn systemC1(_: View<'_, u16>) {}

    let world = World::new();

    let mut group_a = Workload::builder("Group A");
    group_a
        .try_with_system((|world: &World| world.try_run(systemA1), systemA1))
        .unwrap()
        .try_with_system((|world: &World| world.try_run(systemA2), systemA2))
        .unwrap();
    let mut group_b = Workload::builder("Group B");
    group_b
        .try_with_system((|world: &World| world.try_run(systemB1), systemB1))
        .unwrap();
    let mut group_c = Workload::builder("Group C");
    group_c
        .try_with_system((|world: &World| world.try_run(systemC1), systemC1))
        .unwrap();

    let combined_info = Workload::builder("Combined")
        .append(&mut group_a)
        .append(&mut group_b)
        .append(&mut group_c)
        .add_to_world_with_info(&world)
        .unwrap();

    assert_eq!(
        combined_info.batch_info.len(),
        2,
        "systemB1 should move into second batch with systemA2, and systemC1 should move into first batch witch system A1"
    );
    let scheduler = world.scheduler.try_borrow_mut().unwrap();
    assert_eq!(scheduler.systems.len(), 4);
    assert_eq!(scheduler.workloads.len(), 1);
    assert_eq!(
        scheduler.workloads.get("Combined"),
        Some(&vec![vec![0, 3], vec![1, 2]])
    );
    assert_eq!(scheduler.default, "Combined");
}
```

```
failures:

---- world::scheduler::builder::append_ensures_multiple_batches_can_be_optimized_over stdout ----
thread 'world::scheduler::builder::append_ensures_multiple_batches_can_be_optimized_over' panicked at 'assertion failed: `(left == right)`
  left: `Some([[0], [1, 2, 3]])`,
 right: `Some([[0, 3], [1, 2]])`', src/world/scheduler/builder.rs:896:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    world::scheduler::builder::append_ensures_multiple_batches_can_be_optimized_over
```

Reply:
Here, I guess I was expecting that `systemC1` would be moved as far up the "batch stack" as it can go (which would be in batch 1 since no conflicts occur in any existing batches).

WDYT?

Reply:
This is the intended behavior.
With recent discussions on scheduler in the bevy discord I realized that the current workload's description isn't very good. "as much in parallel as possible" isn't true.

Systems can only be added to the current batch or a new one.
In this case `systemA1` and `systemA2` have to be in different batches. Then `systemB1` and `systemC1` can never reach `systemA1`'s batch.

In case you run a workload without parallelism you'll get exactly what you expect. `systemA1` -> `systemA2` -> `systemB1` -> `systemC1`
And then parallelism is just a bonus.

But now that workloads are made with a builder it could be an option.

Fun fact: I very recently rewrote the function making batches, the previous implementation was more complicated than necessary because part of it could handle systems going up multiple batches.

Reply:
Gotcha, that makes sense. So, to retain that promise, we could have `add_to_world_optimized` entry point as an alternative, maybe?

Reply:
Since it's a builder I was thinking of a separate method:

```rust
Workload::builder("W1")
    .max_parallelism()
    .with_system(system!(sys1))
    .add_to_world(&world);
```

But the final result is the same as a `add_to_world_optimized`.

Also it's not 100% clear if it's a speed up or down.
That's also something I want to check and possibly fix. Parallelism has a cost and if the systems don't do enough work it's probably slower to parallelize them at all. Rayon likely plays a big role in this too.